File: day9/day9b.js
const fs = require('node:fs');
const input = fs.readFileSync(__dirname + '/input.txt', 'utf8').trim();

// Here we keep track of the blocks on the disk, every block has:
// id: id of block, null if free space
// size: size of block
let blocks = [];

// Here we keep track of the block ids encountered, this saves looking 
// them up later
let blockIds = [];

let blockId = 0;
let blockSize = null;
input.split('').forEach(i => {
    const j = Number.parseInt(i, 10);
    if (blockSize === null) {
        blockSize = j;
        return;
    }
    // Add block for filled space
    blocks.push({ id: blockId, size: blockSize });
    // Add block for free space
    blocks.push({ id: null, size: j });
    blockIds.push(blockId);
    blockSize = null;

    // Keep track of our blockIds
    blockId++;
});

// The last block has no empty space character so add it now
if (blockId !== null) {    
    blocks.push({ id: blockId, size: blockSize });
    blockIds.push(blockId);
} else {
    blocks[blocks.length-1].sorted = true;
}

const printBlocks = () => {
    // Remove for pretty picture :D
    return;
    let s = '';
    blocks.forEach((b) => {
        for(let i = 0; i < b.size; i++) {
            if (b.id === null) {
                s+= '.';                
            } else {
                s+= b.id;
            }
        }
    })
    console.log(s);
}

// Reverse block ids as we need to move items from high to low
blockIds.reverse();

blockIds.forEach((blockId) => {
    const filledBlockIdx = blocks.findIndex(({id}) => id === blockId);
    const filledBlock = blocks[filledBlockIdx];

    // Search for an empty space before the block index with enough space
    const firstEmptyBlockIdx = blocks.findIndex(({ id, size }) => id === null && size >= filledBlock.size);
    if (firstEmptyBlockIdx === -1 || firstEmptyBlockIdx > filledBlockIdx) {
        // No space, move to next block id
        return;
    }

    const freeBlock = blocks[firstEmptyBlockIdx];

    // scenario 1 -> free block size is same as filled block
    if (freeBlock.size === filledBlock.size) {
        // Move block id and reset block id of last block to null
        freeBlock.id = filledBlock.id;
        filledBlock.id = null;

    // scenario 2 -> free block is larger then filled block}
    } else if (freeBlock.size > filledBlock.size) {
        const remainingFree = freeBlock.size - filledBlock.size;

        // move block id and size from filled block
        freeBlock.id = filledBlock.id;
        freeBlock.size = filledBlock.size;
        filledBlock.id = null;

        // create new free block after it
        blocks.splice(firstEmptyBlockIdx+1, 0, {
            id: null,
            size: remainingFree
        });        
    }

    printBlocks();
})

printBlocks();

let sum = 0;
let idx = 0;
blocks.forEach(({ id, size }) => {
    if (id === null) {
        idx += size;
        return;
    }
    for (let i = 0; i < size; i++) {
        sum += id * idx;
        idx++;
    }
});
console.log(sum);

File: day9_cpp/day9b.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>

using namespace std;

// #define IS_DEBUG
#define FREE_BLOCK_ID -1

struct diskBlock
{
    long blockId;
    char blockSize;

    // Doubly linked-list
    diskBlock* next;
    diskBlock* prev;
};

#ifdef IS_DEBUG
    void printBlocks(diskBlock* pFirstBlock)
    {
        diskBlock* pBlock = pFirstBlock;
        while (pBlock != NULL)
        {
            for (char i = 0; i < pBlock->blockSize; i++) {
                if (pBlock->blockId == FREE_BLOCK_ID) {
                    cout << ".";
                } else {
                    cout << pBlock->blockId;
                }
            }

            pBlock = pBlock->next;
        }
        cout << std::endl;
    }
#endif

diskBlock* findBlockById(diskBlock* pFirstBlock, long blockId)
{
    diskBlock* pBlock = pFirstBlock;
    while (pBlock != NULL)
    {
        if (pBlock->blockId == blockId) {
            return pBlock;
        }
        pBlock = pBlock->next;
    }
    return NULL;
}

diskBlock* findFreeBlock(diskBlock* pFirstBlock, char minSize, long beforeBlockId)
{
    diskBlock* pBlock = pFirstBlock;
    while (pBlock != NULL)
    {
        if (pBlock->blockId == beforeBlockId) {
            return NULL;
        }

        if (
            pBlock->blockId == FREE_BLOCK_ID && 
            pBlock->blockSize >= minSize
        ) {
            return pBlock;
        }

        pBlock = pBlock->next;
    }
    return NULL;
}

diskBlock* insertBlockAfter(diskBlock* pBlock) {
    struct diskBlock *pNewBlock = (struct diskBlock*)malloc(sizeof(struct diskBlock));
    pNewBlock->prev = pBlock;
    pNewBlock->next = pBlock->next;
    if (pBlock->next != NULL) {
        pBlock->next->prev = pNewBlock;
    }
    pBlock->next = pNewBlock;

    return pNewBlock;
}

int main()
{
    #ifdef IS_DEBUG
        std::ifstream fileStream("./test_input.txt");
    #else
        std::ifstream fileStream("./input.txt");
    #endif
    char inputChar = 0;
    
    diskBlock* pFirstBlock = NULL;   
    diskBlock* pLastReadBlock = NULL;

    bool isBlock = true;
    long blockId = 0;
    long lastBlockId = 0;

    while(fileStream.get(inputChar)) {
        char blockSize = inputChar - '0';
        struct diskBlock *b = (struct diskBlock*)malloc(sizeof(struct diskBlock));

        if (isBlock) {            
            b->blockId = blockId;
            b->blockSize = blockSize;
            b->prev = pLastReadBlock;
            b->next = NULL;
            lastBlockId = blockId;
            blockId++;            
        } else {
            b->blockId = FREE_BLOCK_ID;
            b->blockSize = blockSize;
            b->prev = pLastReadBlock;
            b->next = NULL;
        }

        if (pFirstBlock == NULL) {
            pFirstBlock = b;
        }

        if (pLastReadBlock != NULL) {
            pLastReadBlock->next = b;
        }

        pLastReadBlock = b;
        isBlock = !isBlock;
    }

    #ifdef IS_DEBUG
        printBlocks(pFirstBlock);
    #endif

    for (long testBlockId = lastBlockId; testBlockId >= 0; testBlockId--) {
        diskBlock* pBlockById = findBlockById(pFirstBlock, testBlockId);

        // Search for an empty space before the block index with enough space
        diskBlock* pFirstEmptyBlock = findFreeBlock(pFirstBlock, pBlockById->blockSize, pBlockById->blockId);
        if (pFirstEmptyBlock == NULL) {
            continue;
        }

        // scenario 1 -> free block size is same as filled block
        if (pFirstEmptyBlock->blockSize == pBlockById->blockSize) {
            // Move block id and reset block id of last block to null
            pFirstEmptyBlock->blockId = pBlockById->blockId;
            pBlockById->blockId = FREE_BLOCK_ID;

        // scenario 2 -> free block is larger then filled block}
        } else if (pFirstEmptyBlock->blockSize > pBlockById->blockSize) {
            char remainingFree = pFirstEmptyBlock->blockSize - pBlockById->blockSize;

            // move block id and size from filled block
            pFirstEmptyBlock->blockId = pBlockById->blockId;
            pFirstEmptyBlock->blockSize = pBlockById->blockSize;
            pBlockById->blockId = FREE_BLOCK_ID;

            diskBlock* pNewBlock = insertBlockAfter(pFirstEmptyBlock);
            pNewBlock->blockId = FREE_BLOCK_ID;
            pNewBlock->blockSize = remainingFree;
        }

        #ifdef IS_DEBUG
            printBlocks(pFirstBlock);
        #endif
    }

    #ifdef IS_DEBUG
        printBlocks(pFirstBlock);
    #endif

    long sum = 0;
    long idx = 0;
    diskBlock* pBlock = pFirstBlock;
    while (pBlock != NULL) {
        if (pBlock->blockId == FREE_BLOCK_ID) {
            idx += pBlock->blockSize;
        } else {
            for (char i = 0; i < pBlock->blockSize; i++) {
                sum += pBlock->blockId * idx;    
                idx++;
            }
        }
        pBlock = pBlock->next;
    } 
    
    cout << sum << std::endl;

    return 0;
}